### ISSUE TITLE
Investigation: Predicated Rendering

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#551]
From our [conversation](https://github.com/gpuweb/gpuweb/issues/72) about occlusion queries.

## Motivation

One of the natural use cases for using occlusion queries is in conjunction with predicated rendering. One can achieve object culling by rendering the bounding box of some object to an offscreen render target using some trivial shader, and then by using occlusion queries to determine whether the object actually ends up hitting the screen. If it does, then the object can be rendered like normal, but if it doesn't, then the object can be culled.

The problem is that the CPU shouldn't have to wait for the results of the occlusion query to know whether to issue the appropriate draw calls, because that can cause 2-3 frames of latency, which is too late. Instead, predicated rendering allows the GPU to turn a draw call into a no-op based on whether or not a buffer has a nonzero value in a particular place. The occlusion query can be hooked up to populate this value.

## [Direct3D 12](https://docs.microsoft.com/en-us/windows/win32/direct3d12/predication)

There's a command that can be added to the command list: [`ID3DGraphicsCommandList::SetPredication()`](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12graphicscommandlist-setpredication). It takes a buffer and an offset into the buffer. After issuing this command, future draw calls will consult with that 32-bit word in the buffer to determine if they should be turned into no-ops. There's a new resource state: `D3D12_RESOURCE_STATE_PREDICATION`, which the buffer needs to be in (you can transition it to the state using [ID3D12GraphicsCommandList::ResourceBarrier()](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12graphicscommandlist-resourcebarrier)).

But it's not just draw calls. Other operations are predicated too:
- DrawInstanced
- DrawIndexedInstanced
- Dispatch
- CopyTextureRegion
- CopyBufferRegion
- CopyResource
- CopyTiles
- ResolveSubresource
- ClearDepthStencilView
- ClearRenderTargetView
- ClearUnorderedAccessViewUint
- ClearUnorderedAccessViewFloat
- ExecuteIndirect

## [Vulkan](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#drawing-conditional-rendering) - [VK_EXT_conditional_rendering](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_EXT_conditional_rendering)

GPUInfo [says](https://vulkan.gpuinfo.org/displayextension.php?name=VK_EXT_conditional_rendering) 31% Windows, 54% Linux, and 1% Android.

Works very similarly to D3D12. There's a new command you can issue in a command buffer: [vkCmdBeginConditionalRenderingEXT](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCmdBeginConditionalRenderingEXT.html)(), and it takes a buffer and an offset just like D3D12 does. And, like D3D12, there are new flags values that this buffer needs to be [`vkCmdPipelineBarrier()`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCmdPipelineBarrier.html)-ed to: [`VK_ACCESS_CONDITIONAL_RENDERING_READ_BIT_EXT`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccessFlagBits.html), [`VK_BUFFER_USAGE_CONDITIONAL_RENDERING_BIT_EXT`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkBufferUsageFlagBits.html), and [`VK_PIPELINE_STAGE_CONDITIONAL_RENDERING_BIT_EXT`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPipelineStageFlagBits.html).

These operations are conditional:
- drawing commands
- dispatching commands
- [`vkCmdClearAttachments()`](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCmdClearAttachments.html)

## Metal

Metal doesn't have a direct analogue to the other APIs. Instead, it can be achieved two different ways:

- The `indirect` flavor of [draw commands](https://developer.apple.com/documentation/metal/mtlrendercommandencoder). The [`vertexCount`](https://developer.apple.com/documentation/metal/mtldrawprimitivesindirectarguments/1515487-vertexcount) and [`instanceCount`](https://developer.apple.com/documentation/metal/mtldrawprimitivesindirectarguments/1515739-instancecount) fields, as present inside the buffer, can be set to 0, which effectively no-ops the draw call.
- Using [Indirect Command Buffers](https://developer.apple.com/documentation/metal/indirect_command_buffers). These can be [recorded on the GPU](https://developer.apple.com/documentation/metal/indirect_command_buffers/encoding_indirect_command_buffers_on_the_gpu), which means the draw calls themselves can be omitted from the recording. See the [related issue](https://github.com/gpuweb/gpuweb/issues/431) for more discussion.

Neither of these options is available on all Metal devices. The [Metal Feature Set Tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf) contain additional information.

**Comment by litherum** [comment]
We have a couple paths we could take in WebGPU:
1. Take the Metal approach. Say that indirect commands or ExecuteIndirect is enough. Don't add anything beyond those.
2. Add a `setPredication()` command inside render passes and compute passes. Implement it on Metal by emulating it using indirect draws or indirect command buffers. The emulation would require overhead of an amount of GPU buffer memory proportional to the number of draw calls, which is unfortunate. Also, implementation gets tricky when the application itself issues indirect draw calls - we would have to run a compute shader before the render pass to zero-out the relevant fields inside the indirect buffer.

**Comment by kainino0x** [comment]
> Neither of these options is available on all Metal devices. The [Metal Feature Set Tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf) contain additional information.

We did already decide to include indirect draw/dispatch. Although I'm not 100% sure we had the correct info at the time? See this comment and the one above it:
https://github.com/gpuweb/gpuweb/pull/317#issuecomment-505696954
It seems(?) like I thought at the time that all iOS supported it. However, the current feature set tables say it becomes available in MTLGPUFamilyApple3 (Apple A9). But based on the comment just above that, it may just be that we decided that pre-A9 doesn't need to be supported. (I know there was discussion of what hardware was the first we needed to support, but I don't remember which it was.)

**Comment by litherum** [comment]
> We did already decide to include indirect draw/dispatch.

Yes, which was the correct decision. My comment in this thread “ Neither of these options is available on all Metal devices“ was misleading; indirect draws/dispatches are supported on all relevant Metal devices.

**Comment by litherum** [comment]
> Take the Metal approach. Say that indirect commands or ExecuteIndirect is enough.

Because of our buffer uploading rules, this may actually be tricky for a web author to implement themselves right now. We’ll see how the various new buffer upload proposals fare. If they allow an author to append data to a buffer for the current frame performantly, then I would favor path 1.

**Comment by Richard-Yunchao** [comment]
@litherum , thanks for your investigation. I am not sure whether I understood you proposal correctly. But it looks to me that path 1 means that web devs need to do this by themselves in their web apps: 1)use occlusion query in a frame, 2) read/resolve the query data into a GPUBuffer, 3) issue an indirect draw/dispatch, and set indirect buffer accordingly (for example, set instanceCount to 0 if occlusion sample in GPUBuffer is zero). For path 2, it means that we add setPredication API to the WebGPU spec, and web devs need to do nothing. And WebGPU implementers do all of these steps under the hood on Metal. For D3D and Metal, we can call corresponding API at native side directly. 

Another big difference between path 1 and path 2 is the support on devices just as you mentioned.

BTW, [`ID3DGraphicsCommandList::SetPredication()`](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12graphicscommandlist-setpredication) in D3D12 says that it can't be supported in direct command buffer, it should be in a bundle (indirect command buffer).

**Comment by litherum** [comment]
@Richard-Yunchao Yes, that's correct. If authors can already achieve predicated rendering themselves using existing facilities, then that fact significantly weakens the motivation to add it to the browser. Given that the existing facilities work everywhere, on all relevant devices, and even outside of bundles, then option 1 seems fairly appealing.

**Comment by damyanp** [comment]
The comment in the docs about bundles vs direct command lists is clarifying that a bundle executed with ExecuteBundle will inherit the predication state from the command list it is executed from.  You cannot call SetPredication while recording a bundle (but, I would strongly encourage any WebGPU implementation considering using bundles to reconsider their plans).

Explicit predication has some advantages:
* it's easier to program with
* you don't need to over-allocate a buffer big enough for all potential draw calls
* you can predicate indirect draw & dispatch as well as copies, clears and resolves
* drivers can potentially optimize based on predication, for example, avoiding expensive pipeline changes.

**Comment by magcius** [comment]
Predicated rendering tends to be a bigger issue for tilers to support, since could end up synchronizing on the results of other tiles, which can be really expensive.

**Comment by damyanp** [comment]
I think that this would probably also be true of anything that builds up an argument buffer to try and emulate predication using ExecuteIndirect.

**Comment by alecazam** [comment]
It would be nice in Metal if the boolean/counting occlusion query in Metal also present at the same time as Indirect Draws, could set a 4 byte field to 0 (f.e. the primitive count) only if not visible.  But currently Metal occlusions only write to 8 bytes.  The Indirect Draw is all made of 4 byte fields.  This then requires a compute shader just to clobber the numPrimitives field on the indirect draw call.  Also one wouldn't want 1 (boolean) or the count (# pixels) to be copied into the numPrimitives field.  This would be a 4-byte only write zero case for occlusion, and the buffer index/offset would reflect the field of the draw structure.

**Comment by alecazam** [comment]
Even easier that anything presented here is to assume instancing.  Don't try to knock out an ICB or IndirectDraw since that's only a single draw.  Use the Metal occlusion query (or geometry shader on other hw), and overwrite some piece of data in the instance MTLBuffer with the occlusion result.   Then in the vertex shader vertkill all the data if that written value is zero.  One also has a boolean/counting data although so limited material LOD would be possible in the vertex shader.

An API to handling instanced boxes would be nice where occlusion results are written to offsets for multiple boxes using a single call. The current APIs require submitting a draw for each box(es) associated with a single instance.

**Comment by alecazam** [comment]
I feel like occlusion queries are quite limitin for occlusion.  Sampling from a hierarchical z with an O(1) gather operation is a lot more flexible.   This can render all test boxes in one draw call of point primitives, isn't limited by occlusion counts, doesn't lock down depth/raster, supports instancing and indirect draws too.   AC/UE5 takes this further with two pass occlusion.   Vert kill off the occlusion buffer is needed where predicated rendering isn't available.
